[Telemetry]: Add Telemetry Infrastructure + Module Interface + Prometheus Implementation

derrandz opened the pull request:
## Description
##### PR 1/3 for `milestone/v1-telemetry`.


1. Telemetry interface, module and required infrastructure tooling to support it


Notice that we are merging against `milestone/v1-telemetry`. When all telemetry PRs are merged, we will merge `milestone/v1-telemetry` into main.

## Changes

- [x] Added telemetry services to docker-compose
   - [x] loki
   - [x] grafana
   - [x] victoria-metrics
- [x] Defined a telemetry module interface
- [x] Implemented the telemetry module interface for Prometheus
- [x] Implemented a NOOP telemetry module for when telemetry is not desired.

## How to test?

1. Run your stack
```
$ make compose_and_watch
```

3. Check if the infrastructure is up by opening:
  1. [Grafana locally](http://localhost:3000/)
  2. [Victoria Metrics UI](http://localhost:8428/vmui/)
  4. Check to see if Victoria Metrics detects the prometheus endpoint by checking [targets](http://localhost:8428/targets). You should see a section named: `pocket-nodes (4/4)`, with the prometheus endpoint details and status underneath.

  3. Loki by navigating Grafana to the 'Explore' section, and running the query `{host="desktop-docker"} | logfmt` to see logs popping up.

## Documentation

Documentation is available at the telemetry module's [README.md](https://github.com/pokt-network/pocket/blob/telemetry/add-telemetry-infrastructure/shared/modules/telemetry/README.md)

## Owners
@okdas 
@derrandz

Olshansk committed:
Friday night WIP for RainTree happy path no-tests on top of Pre2P - almost working

Olshansk committed:
LocalNet is running but propagation isn't fully functional yet - only reaching 50% of nodes. Need to add debug logs and fix the happy path

Olshansk committed:
Almost working but not progressing on every height

Olshansk committed:
Validator configuration improvements

Olshansk committed:
WIP

Olshansk committed:
Deleted prototype

Olshansk committed:
Add prototype back

Olshansk committed:
Happy path ACTUALLY working now

Olshansk committed:
Added TransportLayerConn prior to writing tests

Olshansk committed:
Replaced Accept w/ TCP in TransportLayerConn abstraction

Olshansk committed:
Pipes are a difficult abstraction - going to try channels

Olshansk committed:
GOT A TEST WORKING - time to sleep

Olshansk committed:
Enforce the correct number of reads & writes

Olshansk committed:
Fix race condition

Olshansk committed:
Add proper demote logic and improve tests

Olshansk committed:
Small & Large RainTree test working consistenly

Olshansk committed:
Cleanup - self review config changes

Olshansk committed:
Cleanup - replaced 'pipeConn' with 'emptyConn'

Olshansk committed:
Rename RawData to NetworkData

Olshansk committed:
Remove unused function

Olshansk committed:
Cleanup - self review of raintree/network code

Olshansk committed:
Cleanup - self review of raintree test

Olshansk committed:
Cleanup - removed and ignoring network_mock.go from git history

Olshansk committed:
TODO to rename 'compose_and_watch'

Olshansk committed:
Fixed genesis configs tests

Olshansk committed:
Fix LocalNet client broadcast

Olshansk committed:
Fix raintree_test expected num reads & writes

Commit:
implemented the cleanup and redundancy layer happy paths

okdas committed:
add victoria metrics and configure docker SD

okdas committed:
some onboarding changes + relabel configs

derrandz committed:
module: add telemetry module with prometheus metrics sdk

derrandz committed:
feat: track pre2p metrics

derrandz committed:
wip(bug): victoria-metrics fails to connect to nodes on 9000

okdas committed:
listen on all interfaces

derrandz committed:
fix: expose http metrics endpoint on external net interface (docker)

derrandz committed:
add more metrics to track and allow switching telemetry on and off to avoid issues with the client

okdas committed:
add more containers

derrandz committed:
chore: fix missing negation at if statement

derrandz committed:
Merge branch 'pre2p/raintree-metrics' of github.com:pokt-network/pocket into pre2p/raintree-metrics

okdas committed:
add loki

okdas committed:
add common lable between metrics and logs

okdas committed:
add monitoring_start to makefile

derrandz committed:
add missing quotes to expose ports in docker-compose

derrandz committed:
Merge branch 'milestone/v1-prototype' into pre2p/raintree-metrics

derrandz committed:
chore: move to laptop

derrandz committed:
feat:
  - add NOOP telemetry module to enable telemetry-less nodes/clients
  - reduce metric list to two (blockheight + total_received_broadcast_msgs)
  - add exported grafana panel in json format

okdas committed:
fixup configs & cleanup

okdas committed:
Fix scraping config regex

derrandz committed:
reduce validators to 4 to let the blockchain go forward (bug?)

Olshansk committed:
Fix make refresh command

Olshansk committed:
Workaroudns for grafana image

derrandz committed:
change machine, revert this commit

derrandz committed:
Merge branch 'pre2p/raintree-metrics' of github.com:pokt-network/pocket into pre2p/raintree-metrics

derrandz committed:
feat: use logging to track events, add grafana dashbaord to aggregate log events and generate graphs

derrandz committed:
Merge branch 'main' into pre2p/raintree-metrics

derrandz committed:
chore: remove non-related changes

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,108 @@
+timeseriesTelemetry := module.GetBus().GetTelemetryModule().GetTimeSeriesAgent()
+// explore the methods you can use in shared/modules/telemetry_module.go
+
+eventMetricsTelemetry.EmitEvent(
```
Why not enforce `namespace`, `event_name` followed by optional args in `EmitEvent`?

derrandz replied:
Affirmative. +1

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,108 @@
+
+#### Consuming logs on loki
+
+To test this out, track an event in your code, and then go to your Grafana's local setup's link, and to the the explore page.
```
`track an event in your code` -> What does this mean?
`Grafana's local setup's link` -> Specify the config

derrandz replied:
Affirmative. Will expand these two.

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,108 @@
+
+To test this out, track an event in your code, and then go to your Grafana's local setup's link, and to the the explore page.
+
+Run the following LogQL query:
```
+1 if we show a screenshot

derrandz replied:
Affirmative.

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,108 @@
+))
+```
+
+#### Node Configuration
````
This should be at the top

derrandz replied:
Affirmative.

Olshansk reviewed (changes requested):
I'm glad we split this out into multiple PRs. I think it's a really good foundational start.

Re the Description, I would do something like this:

```
PR 1/N for `milestone/v1-telemetry`.

1. Telemetry interface, module and required infrastructure tooling to support it
```

It's a lot shorter, and that ways you can build up the list as future PRs come in.

`Run your stack`

Link to the README on how this is done.

`If the telemetry module has been started fine alongside with the rest of the modules, you should see no issues.`

Either remove this line or explain what issues look like.

Bonus points:
- Update the global READMEs with code paths, links to READMEs, changelogs, etc...

`Documentation is available at ...`

You can add a relative link: https://github.blog/2013-01-31-relative-links-in-markup-files/

derrandz commented on Makefile:
```diff
@@ -36,6 +36,18 @@ go_vet:
+## Runs `go mod vendor` && `go mod tidy`
+	go mod vendor && go mod tidy
+
+.PHONY: build
+## Build Pocket's main entrypoint
+build:
```
ACK.

derrandz commented on Makefile:
```diff
@@ -36,6 +36,18 @@ go_vet:
 
+.PHONY: refresh
+## Removes vendor, installs deps, generates mocks and protobuf files. Perform after a new pull or a branch switch
+refresh:
```
Great! adding this.

derrandz commented on Makefile:
```diff
@@ -197,6 +215,7 @@ protogen_clean:
 ## Generate go structures for all of the protobufs
 protogen_local:
 	$(eval proto_dir = "./shared/types/proto/")
+	$(eval shared_flags = "--experimental_allow_proto3_optional")
```
🤔 I am actually not sure about this line... Might've showed up from a merge with main. I don't recall ever needing to touch proto... Will remove.

derrandz commented on shared/config/config.go:
```diff
@@ -13,16 +13,19 @@ import (
-	Pre2P          *Pre2PConfig          `json:"pre2p"` // TECHDEBT(team): consolidate/replace this with P2P configs depending on next steps
+	UseTelemetry bool `json:"use_telemetry"`
+
+	Pre2P          *Pre2PConfig          `json:"pre2p"` // TODO(derrandz): delete this once P2P is ready.
```
`main` merge mistake, correcting shortly.

derrandz commented on shared/config/config.go:
```diff
@@ -13,16 +13,19 @@ import (
 	PrivateKey cryptoPocket.Ed25519PrivateKey `json:"private_key"`
 
-	Pre2P          *Pre2PConfig          `json:"pre2p"` // TECHDEBT(team): consolidate/replace this with P2P configs depending on next steps
+	UseTelemetry bool `json:"use_telemetry"`
```
Works, and is more semantic. Affirmative 👍🏻

derrandz commented on shared/config/config.go:
```diff
@@ -85,6 +88,11 @@ type PersistenceConfig struct {
 type UtilityConfig struct {
 }
 
+type TelemetryConfig struct {
```
Indeed. I will populate this with sufficient documentation. (_There is documentation for it under the telemetry module's README.md under shared, but it should actually be here as well_)

derrandz commented on app/client/main.go:
```diff
@@ -61,6 +64,16 @@ func main() {
 		log.Fatalf("[ERROR] Failed to create pre2p module: " + err.Error())
 	}
 
+	// INFO: This telemetry module instance will be NOOP
```
Affirmative

derrandz commented on build/config/config2.json:
```diff
@@ -2,15 +2,16 @@
     "protocol": "tcp",
-    "address": "0.0.0.0:8084",
-    "external_ip": "172.18.0.1:8084",
+    "address": "0.0.0.0:8082",
```
Yes, I was extensively running the stack and had no issues relating to configuration.

derrandz commented on build/deployments/docker-compose.yaml:
```diff
@@ -1,5 +1,15 @@
 version: "3.8"
 
+# docker plugin install grafana/loki-docker-driver:latest --alias loki --grant-all-permissions
```
Affirmative.

derrandz commented on build/deployments/docker-compose.yaml:
```diff
@@ -108,11 +126,170 @@ services:
   # environment:
   #   - DEBUG_PORT=7084
 
+  node5.consensus:
```
Agreed, I missed this part.

derrandz commented on build/deployments/docker-compose.yaml:
```diff
@@ -141,6 +318,64 @@ services:
+    restart: unless-stopped
+
+  loki:
+    # logging: *loki-logging
```
@okdas Can we consult you on this? I remember I had issues with the loki driver so I commented it out. Let me know.

okdas replied:
This is not really required, we can keep this commented out in case we ever need to feed loki logs into loki. :) Or remove it, too. I understand there might be some dependencies issues (e.g. driver cannot push logs to loki until loki is available), though I did not have any issues loclaly.

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+
+import "github.com/prometheus/client_golang/prometheus"
+
+type (
```
Affirmative.

---

I am in favor of maintaining consistency. I do this out of personal style and habit from older codebases.

Since we've now uncovered this, we can explicitly say that we use this style from now onwards.:

- `type` keyword for instance, no factoring.

derrandz commented on shared/telemetry/module.go:
```diff
@@ -0,0 +1,14 @@
+
+func Create(cfg *config.Config) (modules.TelemetryModule, error) {
+	if cfg.UseTelemetry {
+		return CreatePromModule(cfg)
```
Affirmative for the first suggestion to add a TODO.
Unclear about the second suggestion to add that in the `config.go`

derrandz commented on shared/telemetry/noop_module.go:
```diff
@@ -0,0 +1,103 @@
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var _ modules.TelemetryModule = &PromModule{}
```
Affirmative

derrandz commented on shared/telemetry/noop_module.go:
```diff
@@ -0,0 +1,103 @@
+var _ modules.TelemetryModule = &PromModule{}
+
+// Prometheus struct
+type NoopModule struct {
```
1. Affirmative
2. Was mockgen conceived to be worked with in production for instance? We might want to inject NOOPs for modules that switch on and off telemetry. If so, we can swap it out for a mock.

derrandz commented on consensus/module.go:
```diff
@@ -200,3 +201,7 @@ func (m *consensusModule) handleHotstuffMessage(msg *typesCons.HotstuffMessage)
 	leaderHandlers[msg.Step](m, msg)
 }
+
+func (m *consensusModule) GetBlockHeight() uint64 {
```
Affirmative

derrandz commented on p2p/pre2p/module.go:
```diff
@@ -108,7 +111,12 @@ func (m *p2pModule) Broadcast(msg *anypb.Any, topic types.PocketTopic) error {
 	log.Println("broadcasting message to network")
-	return m.network.NetworkBroadcast(data)
+
+	if err := m.network.NetworkBroadcast(data); err != nil {
```
It was not an explicit change by myself, but potentially a `main` merge mistake. Will undo.

derrandz commented on p2p/pre2p/module.go:
```diff
@@ -121,7 +129,11 @@ func (m *p2pModule) Send(addr cryptoPocket.Address, msg *anypb.Any, topic types.
 		return err
 	}
 
-	return m.network.NetworkSend(data, addr)
```
These were unintentional and probably the result of a `main` merge conflict resolution mistake. Thanks for catching these. Undoing.

derrandz commented on p2p/pre2p/module.go:
```diff
@@ -23,6 +23,8 @@ import (
 var _ modules.P2PModule = &p2pModule{}
 
 type p2pModule struct {
+	// p2pModule is a telemetryModule, that will use the shared telemetry module in the bus
```
This is legacy, thanks for cleaning up for me.

derrandz commented on shared/bus.go:
```diff
@@ -14,30 +14,78 @@ type bus struct {
+	return bus
+}
+
 func CreateBus(
```
Affirmative

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+		// Counters
+
+		// Register a counter by name
+		RegisterCounter(name string, description string)
```
No, I hear you. Now that I see it I would be biased to this approach, and perhaps is more go-idiomatic, as it feels quite familiar.

Affirmative, incorporating this change.

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+
+		// Register a gauge by name
+		RegisterGauge(name string, description string)
+		// Set sets the Gauge to an arbitrary value.
```
Affirtmative.

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+		// Register a gauge by name
+		RegisterGauge(name string, description string)
+		// Set sets the Gauge to an arbitrary value.
+		SetGauge(name string, value float64) prometheus.Gauge
```
This is somewhat of a historic thing that remained. I initially was interesting in  chaining telemetry calls to make use of `WithLabelValues` and `WithLabels` calls. So we could do away with this change as it's not used anywhere at the moment, but we can make use of it if we keepit, and it would help us do something like the following:
```
m.SetGauge("x", 0).WithLabelValues(..).Add(10)
```

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+		RegisterGauge(name string, description string)
+		// Set sets the Gauge to an arbitrary value.
+		SetGauge(name string, value float64) prometheus.Gauge
+		// Inc increments the Gauge by 1. Use Add to increment it by arbitrary
```
Affirmative

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+		// Dec decrements the Gauge by 1. Use Sub to decrement it by arbitrary
+		// values.
+		DecGauge(name string) prometheus.Gauge
+		// Add adds the given value to the Gauge. (The value can be negative,
```
Affirmative.

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+		// Register a gauge vector by name and provide labels
+		RegisterGaugeVector(namespace, module, name, description string, labels []string)
+		// Retrieve a gauge vector by name
+		GetGaugeVec(name string) *prometheus.GaugeVec
```
Affirmative.

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,54 @@
+
+	// Interface for the event metrics agent (relies on logging ftm)
+	EventMetricsAgent interface {
+		EmitEvent(...interface{})
```
Interesting! Affirmative.

derrandz commented on docs/README.md:
```diff
@@ -1,25 +0,0 @@
-<div align="center">
```
I might need your help on that, I was confused when I saw these changes too. I made sure the merge wen't fine and that conflicts were resolved, not sure exactly how we got those here...

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// INFO: At the moment, we are using loki to push log lines per event emission, and then we aggregate these log lines (count them, avg, etc) in grafana.
+// Reliance on logs for event metrics was a temporary solution, and will be removed in the future in favor of more thorough event metrics tooling.
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+
+func (p *PromModule) RegisterCounter(name string, description string) {
+	if _, exists := p.counters[name]; exists {
+		return
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+
+func (p *PromModule) IncCounter(name string) {
+	if _, exists := p.counters[name]; !exists {
+		return
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+	})
+}
+
+// Set sets the Gauge to an arbitrary value.
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+	return gg
+}
+
+// IncGauge(name string)
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+// Set sets the Gauge to an arbitrary value.
+func (p *PromModule) SetGauge(name string, value float64) prometheus.Gauge {
+	if _, exists := p.gauges[name]; !exists {
+		return nil
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,198 @@
+	gg := p.gauges[name]
+	gg.Set(value)
+
+	return gg
```
Yes, I've addressed it in that comment. I had to expose the gauge to make use of its full functionality. (_`WithLabels`, `WithLabelValues` etc.._)

derrandz commented on shared/telemetry/noop_module.go:
```diff
@@ -0,0 +1,103 @@
+	return m
+}
+
+// INFO: At the moment, we are using loki to push log lines per event emission, and then we aggregate these log lines (count them, avg, etc) in grafana.
```
Affirmative

derrandz commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,108 @@
+
+### Event Metrics
+
+In the current implementation, we are recording events through logs. For each unique witnessing of a given event, we will emit a log line with specific formatting to stdout.
```
🤔 I hear you. It's extra-detail. I will add this change.

derrandz committed:
add back removed files

derrandz committed:
add back CONTRIBUTING.md

derrandz committed:
implement requested changes

derrandz commented:
Hey @Olshansk Thank you for your thorough review.

I've addressed all of the points in question, I still have to add in there a few screenshots for the documentation, but have left TODOs for that.

Let me know if you can approve this PR.

derrandz committed:
fix any keyword syntax and upgrade docker image to 1.18

okdas commented on build/deployments/telemetry/dashboards/victoriametrics.json:
```diff
@@ -0,0 +1,4263 @@
+{
```
All json files in `[build/deployments/telemetry/dashboards](https://github.com/pokt-network/pocket/pull/95/files#diff-322ff7dd9d4f70d4869671fec0d70abd23070b1ce0bc5ec7a274850fc4fa8bd3)` directory create dashboards in Grafana, and this is just one of them. I'd keep it for local victoria metrics troubleshooting purposes.

okdas commented on build/deployments/telemetry/victoria-metrics/scrape.yml:
```diff
@@ -0,0 +1,30 @@
+# You can follow prometheus config syntax here.
```
```suggestion
# You can follow prometheus config syntax in this file.
```

@Olshansk correct - in this file.

okdas committed:
fix loginless grafana

okdas committed:
rm build/granafa/provisioning/datasources/datasource.yaml

okdas commented on build/deployments/telemetry/victoria-metrics/scrape.yml:
```diff
@@ -0,0 +1,30 @@
+      - source_labels:
+          [__meta_docker_container_label_com_docker_compose_service]
+        target_label: instance
+        replacement: "${1}"
```
@Olshansk it is not, I removed it.

okdas committed:
This config is actually also not in use.

okdas reviewed (approved):
Loki/Metrics/Grafana LGTM - I changed the environment variables a bit b/c with previous setup Grafana was still asking a login. Now, on fresh installation, it won't anymore. Also removed unnecessary files/configs.

<img src="https://media1.giphy.com/media/73Y6EEJqbyGovenlLE/giphy.gif"/>

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+
+At the moment, we are using two types of metrics:
+
+- Time series metrics: collected periodically and are stored in a database.
```
NIT: Make `Time Series metics` bold. DItto for event series metric and other headers in bullet points below.

Olshansk replied:
Also, maybe this is clearer: "Metrics collected at regular intervals and stored in a database"

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+At the moment, we are using two types of metrics:
+
+- Time series metrics: collected periodically and are stored in a database.
+- Event series metrics: events that occur and are tracked and stored in a database.
```
`events that occur and are tracked and stored in a database.` -> `event-driven (i.e. not time-driven) metrics that are stored in a databse`

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+
+At the moment, we are using:
+- Prometheus: for timeseries
+- Plain Logs: for event metrics (_Might be subsituted in the future with an events database_)
```
Instead of `MIGHT`, add an appropriate `TODO` or `INVESTIGATE` or future work flag

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+
+## Node Configuration
+
+It is necessary to provide a telemetry configuration to your node:
```
to your node in `config.json`:

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+  }
+```
+
+`enable_telemetry`: is a boolean json entry defined at the root of the document that tells the node whether to use the telemetry module or use a NOOP version.
````
`is a boolean json entry defined at the root of the document that tells the node whether to use the telemetry module or use a NOOP version` -> `configures node to expose telemetry if true`?

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+
+`enable_telemetry`: is a boolean json entry defined at the root of the document that tells the node whether to use the telemetry module or use a NOOP version.
+`address`: is the prometheus server's address that the telemetry module will listen on.
+`endpoint`: the scraping endpoint that prometheus exposes through the telemetry module.
```
Why are we using the word `scraping` here?

derrandz replied:
Telemetry nomenclature. If you take a peak at the `victoria-metrics` configuration, you will find that the configuration key to hook it up to prometheus (_for instance_) is called: `scrape:`...

I've changed this comment to avoid the confusion.

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+## Time Series Metrics
+
+If you aren't familiar with time series metrics that Prometheus offers, please check out [Prometheus Metrics](https://prometheus.io/docs/concepts/metric_types/)
+
```
Extra newline

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+
+## Time Series Metrics
+
+If you aren't familiar with time series metrics that Prometheus offers, please check out [Prometheus Metrics](https://prometheus.io/docs/concepts/metric_types/)
```
`If you are not familiar with the time-series concepts related to Prometheus, you can review [Prometheus Metrics](https://prometheus.io/docs/concepts/metric_types/).`

Olshansk commented on shared/telemetry/README.md:
```diff
@@ -0,0 +1,121 @@
+We use Gauges to keep track of:
+
+- Blockheight
+- Nodes Online
```
- Number of Nodes online?

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+In your module, make sure you have access to the bus, then use the metrics you need as follows:
+```go
+
+timeseriesTelemetry := module.GetBus().GetTelemetryModule().GetTimeSeriesAgent()
````
I think this is another copy-pasta error.

`eventMetricsTelemetry` is not declared anywhere

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+)
+```
+
+### Consuming logs on loki
````
in loki

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+```
+{host="desktop-docker"} |= "[EVENT] your_namespace your_event" | pattern `<datetime> <_> <time> <type> <event_name> <any> <aditional> <whitespaced> <items>` | logfmt
+```
+TODO(derrandz): add screenshot in here
````
Could you add it as part of this PR?

Even better, a gif

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+To launch and start using Grafana, do the following:
+
+1. Spin up the stack
+```
````
You can add `bash` to code snippets like this so there's additional formatting for the reader.

Olshansk commented on shared/telemetry/README.md:
````diff
@@ -0,0 +1,121 @@
+```
+
+2. Wait a few seconds and then visit: `https://localhost:3000`.
+3. Voila! You are there. You can browse existing dashbaords by (x)s (TODO(derrandz): Add screenshot)
````
NIT: Screenshots would help me (the reader and reviewer) understand how to run it myself.

Olshansk commented on shared/telemetry/noop_telemetry_module.go:
```diff
@@ -0,0 +1,107 @@
+	NOOP()
+}
+
+// Set sets the Gauge to an arbitrary value.
```
You kept the comments for some functions but not others.

1. Stay consistent.
2. It's confusing to have these comments in place for the NOOP module.

derrandz commented on shared/config/config.go:
```diff
@@ -85,6 +88,11 @@ type PersistenceConfig struct {
 }
 
+type TelemetryConfig struct {
+	Address  string // The address that the telemetry module will use to listen for scrape requests (e.g: 0.0.0.0:9000 for proemetheus)
```
It's common in observability nomenclature to use the term 'scrape', but you are right, it does make me think that way too. I've changed it to `metrics pulling`

derrandz commented on shared/modules/telemetry_module.go:
```diff
@@ -0,0 +1,53 @@
+	// Counters
+
+	// Registers a counter by name
+	CounterRegister(name string, description string)
```
No. At the moment, if you try to re-register anything you've already registered, we ignore the fact and carry on cuz the job was done.

I see what you are saying, but this can be looked at as an improvement. When we are happy with the base functionality, we can revise code behavior however we see fit.

derrandz commented on shared/telemetry/noop_telemetry_module.go:
```diff
@@ -0,0 +1,107 @@
+package telemetry
```
Affirmative

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,199 @@
+	_ modules.TimeSeriesAgent   = &PromModule{}
+)
+
+type PromModule struct {
```
I've eventually renamed it to `PrometheusTelemetryModule`.

derrandz commented on shared/telemetry/prometheus_module.go:
```diff
@@ -0,0 +1,199 @@
+	return m.bus
+}
+
+// Event Metrics functionality implementation
```
I meant to write `EventMetricsAgent interface implementation`

derrandz committed:
add requested changes

derrandz committed:
Merge branch 'telemetry/add-telemetry-infrastructure' of github.com:pokt-network/pocket into telemetry/add-telemetry-infrastructure

derrandz commented:
@Olshansk Thank you for your review and for the level of detail to which you go.

For everything that I agree with, I have simply pushed changes and resolved.
For the other comments I've changed and still require your review on them, I've kept unresolved.

For the telemetry module creation issue, I've produced the following: https://github.com/pokt-network/pocket/issues/99

Please let me know.

Olshansk committed:
Update comment in main.go

Olshansk committed:
Remove mocks/network_mock.go from git history

Olshansk committed:
Remove redundancy in 'make refresh' target

Olshansk committed:
Remove unnecessary panicOnMissingMod fn

Olshansk committed:
Simplify code with maybeSetModuleBus

Olshansk committed:
Simplify TelemetryConfig comments

Olshansk committed:
Some code cleanup to the telemetry module interfaces and implementations

Olshansk committed:
Updated the documentation

Olshansk committed:
Increase the image version of the docker client

Olshansk reviewed (approved):
Thanks @derrandz!

There were a few minor nits I fixed and updated myself. For example:
* Random typos
* Changes / fixes that were done in one place but not in another

I also updated some of the documentation as I was following it while getting my local stack setup.

PTAL (Please Take A Look) at the changes I pushed and let me know if you see any issues with it.

🚢 **Other than that, LGTM!** 🚢

Olshansk commented:
**IMPORTANT:** Please make sure to **squash & merge** this commit and include the description from GitHub issues at the body of the squash commit.
